state/backups/create.go

```diff
@@ -109,9 +110,9 @@ type builder struct {
 // directories which backup uses as its staging area while building the
 // archive.  It also creates the archive
 // (temp root, tarball root, DB dumpdir), along with any error.
-func newBuilder(filesToBackUp []string, db DBDumper) (b *builder, err error) {
+func newBuilder(backupDir string, filesToBackUp []string, db DBDumper) (b *builder, err error) {
 	// Create the backups workspace root directory.
-	rootDir, err := ioutil.TempDir("", tempPrefix)
+	rootDir, err := ioutil.TempDir(backupDir, tempPrefix)
```

Comment:
here is where backup dir needs to be validated, unless TempDir() fails anyway

Reply:
func TempDir(dir, prefix string) (name string, err error) {
	if dir == "" {
		dir = os.TempDir()
	}

	nconflict := 0
	for i := 0; i < 10000; i++ {
		try := filepath.Join(dir, prefix+nextSuffix())
		err = os.Mkdir(try, 0700)
		if os.IsExist(err) {
			if nconflict++; nconflict > 10 {
				randmu.Lock()
				rand = reseed()
				randmu.Unlock()
			}
			continue
		}
		if os.IsNotExist(err) {
			if _, err := os.Stat(dir); os.IsNotExist(err) {
				return "", err
			}
		}
		if err == nil {
			name = try
		}
		break
	}
	return}

In the ioutil.TempDir, if the provided folder doesn't exist. The function fails. and we catch that error on line 115.